File: articles/self-determination-theory-positive-psychology/cpp/sdt_indices.cpp
#include <iostream>

struct SDTIndicators {
    double autonomy_support;
    double competence_support;
    double relatedness_support;
    double need_frustration;
    double controlling_pressure;
    double autonomous_motivation;
    double controlled_motivation;
    double internalization;
    double wellbeing_score;
    double vitality;
    double stress_load;
    double climate_quality;
};

double need_support_index(const SDTIndicators& x) {
    return (x.autonomy_support + x.competence_support + x.relatedness_support) / 3.0;
}

double need_balance_index(const SDTIndicators& x) {
    return need_support_index(x) - x.need_frustration - x.controlling_pressure;
}

double motivational_quality_index(const SDTIndicators& x) {
    return x.autonomous_motivation + x.internalization - x.controlled_motivation;
}

double net_sdt_wellbeing_index(const SDTIndicators& x) {
    return x.wellbeing_score
         + x.vitality
         + x.autonomous_motivation
         + x.internalization
         + need_support_index(x)
         - x.controlled_motivation
         - x.stress_load
         - x.need_frustration
         - x.controlling_pressure;
}

int main() {
    SDTIndicators example{7.4, 7.3, 7.5, 2.0, 2.4, 7.3, 2.8, 6.9, 7.4, 7.3, 2.9, 7.6};

    std::cout << "Need support index: " << need_support_index(example) << "\n";
    std::cout << "Need balance index: " << need_balance_index(example) << "\n";
    std::cout << "Motivational quality index: " << motivational_quality_index(example) << "\n";
    std::cout << "Net SDT wellbeing index: " << net_sdt_wellbeing_index(example) << "\n";

    return 0;
}

File: articles/self-determination-theory-positive-psychology/go/sdt_indices.go
package main

import "fmt"

type SDTIndicators struct {
	AutonomySupport      float64
	CompetenceSupport    float64
	RelatednessSupport   float64
	NeedFrustration      float64
	ControllingPressure  float64
	AutonomousMotivation float64
	ControlledMotivation float64
	Internalization      float64
	WellbeingScore       float64
	Vitality             float64
	StressLoad           float64
	ClimateQuality       float64
}

func NeedSupportIndex(x SDTIndicators) float64 {
	return (x.AutonomySupport + x.CompetenceSupport + x.RelatednessSupport) / 3.0
}

func NeedBalanceIndex(x SDTIndicators) float64 {
	return NeedSupportIndex(x) - x.NeedFrustration - x.ControllingPressure
}

func MotivationalQualityIndex(x SDTIndicators) float64 {
	return x.AutonomousMotivation + x.Internalization - x.ControlledMotivation
}

func NetSDTWellbeingIndex(x SDTIndicators) float64 {
	return x.WellbeingScore +
		x.Vitality +
		x.AutonomousMotivation +
		x.Internalization +
		NeedSupportIndex(x) -
		x.ControlledMotivation -
		x.StressLoad -
		x.NeedFrustration -
		x.ControllingPressure
}

func main() {
	example := SDTIndicators{7.4, 7.3, 7.5, 2.0, 2.4, 7.3, 2.8, 6.9, 7.4, 7.3, 2.9, 7.6}
	fmt.Printf("Need support index: %.3f\n", NeedSupportIndex(example))
	fmt.Printf("Need balance index: %.3f\n", NeedBalanceIndex(example))
	fmt.Printf("Motivational quality index: %.3f\n", MotivationalQualityIndex(example))
	fmt.Printf("Net SDT wellbeing index: %.3f\n", NetSDTWellbeingIndex(example))
}
